### PR TITLE
Remove OpenInputFile and CreateOutputFile methods from IHostEnvironment

### PR DESCRIPTION
**Pull request by maryamariyan**
Didn't add new tests since the implementation is just moved from HostEnvironmentBase to HostEnvironmentUtils class.

On a separate PR will remove CreateTempFile which still part of IHostEnvironment interface.

Related to: #1287
cc: @eerhardt 

- [x] There's a descriptive title that will make sense to other developers some time from now. 
- [x] There's associated issues. All PR's should have issue(s) associated - unless a trivial self-evident change such as fixing a typo. You can use the format `Fixes #nnnn` in your description to cause GitHub to automatically close the issue(s) when your PR is merged.
- [x] Your change description explains what the change does, why you chose your approach, and anything else that reviewers should know.
- [ ] You have included any necessary tests in the same PR.

### CODE REVIEW

File: src/Microsoft.ML.Core/Data/IHostEnvironment.cs [hunk 1]
```diff
@@ -23,6 +23,28 @@ public interface IChannelProvider : IExceptionContext
         IPipe<TMessage> StartPipe<TMessage>(string name);
     }
 
+    /// <summary>
+    /// Utility class for IHostEnvironment
+    /// </summary>
+    internal static class HostEnvironmentExtensions
```
Comment (eerhardt) [inline src/Microsoft.ML.Core/Data/IHostEnvironment.cs]: I'm surprised this doesn't need a `[BestFriend]` attribute.  @TomFinley - do you know why?

File: src/Microsoft.ML.Core/Environment/HostEnvironmentBase.cs [hunk 2]
```diff
@@ -460,41 +452,6 @@ protected virtual IProgressChannel StartProgressChannelCore(HostBase host, strin
...
 
-        /// <summary>
-        /// This "opens" the input file handle. The default implementation assumes the file resides in the
-        /// file system and it does not access the physical file until a read stream is opened on the returned
-        /// file handle.
-        /// </summary>
-        protected virtual IFileHandle OpenInputFileCore(IHostEnvironment env, string path)
-        {
-            this.AssertValue(env);
-            this.CheckNonWhiteSpace(path, nameof(path));
```
Comment (eerhardt) [inline src/Microsoft.ML.Core/Environment/HostEnvironmentBase.cs]: Are we missing these checks in the new Utility methods?

**Comment by maryamariyan** [comment]
I was going to create another PR for removing IHostEnvironment.CreateTempFile, but I just pushed it up here to the same PR.

Locally all tests passed for me using `build.cmd -runTests`

Last thing to check will be to look to see if we need the corner cases as mentioned by @eerhardt [above ](https://github.com/dotnet/machinelearning/pull/2539#discussion_r256631145)

File: src/Microsoft.ML.Data/EntryPoints/Cache.cs [hunk 3]
```diff
@@ -34,7 +34,7 @@ public sealed class CacheOutput
         }
 
         [TlcModule.EntryPoint(Name = "Transforms.DataCache", Desc = "Caches using the specified cache option.", UserName = "Cache Data")]
-        public static CacheOutput CacheData(IHostEnvironment env, CacheInput input)
+        public static CacheOutput CacheData(IHostEnvironment env, CacheInput input, IFileHandle fileHandle)
```
Comment (eerhardt) [inline src/Microsoft.ML.Data/EntryPoints/Cache.cs]: Since this is an `EntryPoint` itself, I don't know if we can add parameters to it. I'm not 100% exactly familiar with how EntryPoints work.

@ganik ?

File: src/Microsoft.ML.Core/Environment/HostEnvironmentBase.cs [hunk 4]
```diff
@@ -460,98 +452,6 @@ protected virtual IProgressChannel StartProgressChannelCore(HostBase host, strin
...
-
-            Guid guid = Guid.NewGuid();
-            string path = Path.GetFullPath(Path.Combine(Path.GetTempPath(), prefix + guid.ToString() + suffix));
-            return new SimpleFileHandle(env, path, needsWrite: true, autoDelete: true);
-        }
-
-        /// <summary>
-        /// Returns true if the given string is non-null and contains invalid file name characters.
-        /// </summary>
-        protected virtual bool HasBadFileCharacters(string str = null)
```
Comment (eerhardt) [inline src/Microsoft.ML.Core/Environment/HostEnvironmentBase.cs]: Are we still doing these checks on the new Util method?

File: test/Microsoft.ML.TestFramework/BaseTestBaseline.cs [hunk 5]
```diff
@@ -76,6 +76,8 @@ protected BaseTestBaseline(ITestOutputHelper output) : base(output)
         protected StreamWriter LogWriter;
         private protected ConsoleEnvironment _env;
         protected IHostEnvironment Env => _env;
+        private IFileHandle _fileHandle;
```
Comment (eerhardt) [inline test/Microsoft.ML.TestFramework/BaseTestBaseline.cs]: I wouldn't add this functionality to the base test class just for a single spot. Let's leave this off the base class, and just create and dispose the temp file in the one place that needs it.

File: src/Microsoft.ML.Data/EntryPoints/Cache.cs (hunk 3 above)
Comment (eerhardt) [reply src/Microsoft.ML.Data/EntryPoints/Cache.cs]: Maybe it could work something like this:

The `CacheInput` can take a path to where on the file system to put the cached file. And then if disk caching was specified the `CacheOutput` will have the path to the file. It is then up to the caller to clean the file up when they are done with it.

Thoughts?

---
In reply to: [256896499](https://github.com/dotnet/machinelearning/pull/2539#discussion_r256896499) [](ancestors = 256896499)

**Comment by eerhardt** [comment]
> I was going to create another PR for removing IHostEnvironment.CreateTempFile, but I just pushed it up here to the same PR.

I think it makes sense to split them into separate changes. The first one is pretty trivial and could probably be merged today. The 2nd one is going to take some more work in order to get it merged.

**Comment by maryamariyan** [comment]
Moved the diff for removing CreateTempFile into a PR in my own fork [here](https://github.com/maryamariyan/machinelearning/pull/1).

File: src/Microsoft.ML.Core/Environment/HostEnvironmentBase.cs (hunk 2 above)
Comment (maryamariyan) [reply src/Microsoft.ML.Core/Environment/HostEnvironmentBase.cs]: TODO: 
- [ ] bring back checks to utility methods

File: test/Microsoft.ML.TestFramework/BaseTestBaseline.cs (hunk 5 above)
Comment (maryamariyan) [reply test/Microsoft.ML.TestFramework/BaseTestBaseline.cs]: comment for next PR

File: src/Microsoft.ML.Core/Environment/HostEnvironmentBase.cs (hunk 4 above)
Comment (maryamariyan) [reply src/Microsoft.ML.Core/Environment/HostEnvironmentBase.cs]: comment for next PR

File: src/Microsoft.ML.Data/EntryPoints/Cache.cs (hunk 3 above)
Comment (maryamariyan) [reply src/Microsoft.ML.Data/EntryPoints/Cache.cs]: comment for next PR

**Comment by maryamariyan** [comment]
@dotnet-bot test MachineLearning-CI

**Comment by eerhardt** [comment]
Thanks @maryamariyan!
